cluster-autoscaler/cloudprovider/clusterapi/clusterapi_nodegroup.go

```diff
@@ -203,11 +203,28 @@ func (ng *nodegroup) DecreaseTargetSize(delta int) error {
 		return err
 	}
 
-	if size+delta < len(nodes) {
-		return fmt.Errorf("attempt to delete existing nodes targetSize:%d delta:%d existingNodes: %d",
-			size, delta, len(nodes))
+	// we want to filter out machines that are not nodes (eg failed or pending)
+	// so that the node group size can be set properly. this affects situations
+	// where an instance is created, but cannot become a node due to cloud provider
+	// issues such as quota limitations, and thus the autoscaler needs to properly
+	// set the size of the node group. without this adjustment, the core autoscaler
+	// will become confused about the state of nodes and instances in the clusterapi
+	// provider.
+	actualNodes := 0
+	for _, node := range nodes {
+		if !isPendingMachineProviderID(normalizedProviderID(node.Id)) &&
+			!isFailedMachineProviderID(normalizedProviderID(node.Id)) &&
+			!isDeletingMachineProviderID(normalizedProviderID(node.Id)) {
```

Comment:
Is it not possible to do something like `if node.Status.State == cloudprovider.InstanceDeleting {` instead of mucking with the providerID string and then having to do string in-line mutations in order to extract the providerID bit when we need it?

Reply:
i'll have to investigate that deeper to see if it would work.

i tried to follow the pattern in https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go#L597 which returns a list of provider ids, and not specifically the node objects.

Reply:
i looked at the code a little more, part of the problem here is that we are not filling in all the details for the `Instances` structure that is returned as part of `nodes`. see https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/clusterapi/clusterapi_nodegroup.go#L241-L245

i had thought about improving that part of the function to set the proper state for the Instance, but that is a much larger change than what i'm proposing, and it would also rely on using the provider string mutation method.

Reply:
ACK, let's move forward with a more targeted fix, acknowledging that some of our current difficulty is that our foundational node accounting is incomplete. Pending time/priority, I think addressing that will reduce these edge cases in the long term.

Reply:
thanks, i'm trying to find some time to do more refactoring here. i think the tests could use it, and then it will make it easier to add more testing and structure for the other bits of code.